algorithms/boyer_moore.go

```diff
@@ -0,0 +1,35 @@
+package algorithms
+
+func Search(needle []rune, haystack []rune) int {
+	n := len(needle)
+	if n == 0 {
+		return -1
+	}
+	maxLen := len(haystack)
+	skipTable := skipTable(needle)
+
+	position := n - 1
+	for position < maxLen {
+		i := n - 1
+		for ; i >= 0 && needle[i] == haystack[position]; i, position = i-1, position-1 {
+		}
+		if i == -1 {
+			return position + 1
+		}
+		if shift, exist := skipTable[haystack[position]]; exist {
+			position = position + shift
+		} else {
+			position = position + n
+		}
+	}
+	return -1
+}
+
+func skipTable(needle []rune) map[rune]int {
+	result := make(map[rune]int)
+	n := len(needle)
+	for i := 0; i < n; i++ {
+		result[needle[i]] = n - 1 - i
+	}
```

Comment:
The original approach looked like this:
```GO
	for i := n - 1; i >= 0; i-- {
		_, exists := result[needle[i]]
		if !exists {
			result[needle[i]] = n - 1 - i
		}
	}
```
It could be reduced like above to this:
```GO
	for i := n - 1; i >= 0; i-- {
		if _, exists := result[needle[i]]; !exists {
			result[needle[i]] = n - 1 - i
		}
	}
```
Instead I decided to scrap the existance idea and instead iterate over the whole needle and add the values to the `result` array. Since character may appear multiple times, the iteration had to be reversed and therefor the final value for a character will always be the one from the closest to the end.